### PR TITLE
Reduce the amount of untracked state in TyCtxt

### PR DESCRIPTION
**Pull request by cjgillot**
Access to untracked global state may generate instances of #84970.

The GlobalCtxt contains the lowered HIR, the resolver outputs and interners.
By wrapping the resolver inside a query, we make sure those accesses are properly tracked.
As a no_hash query, all dependent queries essentially become `eval_always`,
what they should have been from the beginning.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -1267,22 +1227,26 @@ impl<'tcx> TyCtxt<'tcx> {
...
             self.hir().def_path(id)
         } else {
-            self.cstore.def_path(id)
+            self.untracked_resolutions.cstore.def_path(id)
         }
     }
 
     /// Returns whether or not the crate with CrateNum 'cnum'
     /// is marked as a private dependency
     pub fn is_private_dep(self, cnum: CrateNum) -> bool {
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/context.rs:1236]: Should this include `untracked` in the name?

Comment (cjgillot) [reply compiler/rustc_middle/src/ty/context.rs:1236]: I added untracked in the name, as a warning to the reader that this value is not tracked by the query system, and may cause silent invalidations.

File: compiler/rustc_middle/src/ty/context.rs [hunk 2]
```diff
@@ -1247,15 +1203,20 @@ impl<'tcx> TyCtxt<'tcx> {
     }
 
     pub fn allocator_kind(self) -> Option<AllocatorKind> {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/context.rs:1205]: Should uses of this method be audited?

Comment (michaelwoerister) [reply compiler/rustc_middle/src/ty/context.rs:1205]: Can we just turn it into a query depending on the new `resolutions` query?

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 3]
```diff
@@ -155,19 +155,22 @@ impl<'hir> Map<'hir> {
 
     #[inline]
     pub fn definitions(&self) -> &'hir Definitions {
-        &self.tcx.definitions
+        // Accessing the definitions is ok, since all its contents are tracked by the query system.
```
Comment (Aaron1011) [inline compiler/rustc_middle/src/hir/map/mod.rs:158]: Can you elaborate this? I'm not sure where that tracking happens.

Comment (cjgillot) [reply compiler/rustc_middle/src/hir/map/mod.rs:158]: Now you ask, I have a doubt this is true for *all* fields in `definitions`. The `LocalDefId <-> DefPath/DefPathHash/DefKey` maps are stable by construction. The stability of the `LocalDefId <-> HirId` maps is related to #85266.
The other fields may not be stable :frowning_face:

File: compiler/rustc_middle/src/ty/context.rs [hunk 4]
```diff
@@ -1310,32 +1264,39 @@ impl<'tcx> TyCtxt<'tcx> {
     }
 
     pub fn metadata_encoding_version(self) -> Vec<u8> {
-        self.cstore.metadata_encoding_version().to_vec()
+        self.untracked_resolutions.cstore.metadata_encoding_version().to_vec()
```
Comment (Aaron1011) [inline compiler/rustc_middle/src/ty/context.rs:1267]: Can you add comments to this method and `encode_metadata` justifying the use of `untracked_resolutions`, like you did in other methods?

File: compiler/rustc_resolve/src/lib.rs [hunk 5]
```diff
@@ -1132,8 +1132,8 @@ impl ResolverAstLowering for Resolver<'_> {
         self.next_node_id()
     }
 
-    fn trait_map(&self) -> &NodeMap<Vec<TraitCandidate>> {
-        &self.trait_map
+    fn trait_map(&mut self) -> NodeMap<Vec<TraitCandidate>> {
+        std::mem::take(&mut self.trait_map)
```
Comment (Aaron1011) [inline compiler/rustc_resolve/src/lib.rs:1136]: Any subsequent calls to `trait_map()` will return an empty `NodeMap`, which seems like a footgun. Could you add a (debug) assertion to verify that this is only called once?

File: compiler/rustc_resolve/src/lib.rs [hunk 6]
```diff
@@ -1133,6 +1135,11 @@ impl ResolverAstLowering for Resolver<'_> {
     }
 
     fn trait_map(&mut self) -> NodeMap<Vec<TraitCandidate>> {
```
Comment (bjorn3) [inline compiler/rustc_resolve/src/lib.rs:1137]:
```suggestion
    fn take_trait_map(&mut self) -> NodeMap<Vec<TraitCandidate>> {
```

**Comment by Aaron1011** [comment]
r=me, assuming that @bjorn3 and @michaelwoerister don't have any other comments.

**Review by bjorn3 — APPROVED** [review]

**Comment by goffrie** [comment]
Seems like this PR [regressed performance quite a bit](https://perf.rust-lang.org/compare.html?start=d2091147b16a1e3e1f9a73b6ee29fbe941c8abce&end=41278062c8399ac02ed281e8b1648b99a36942e6&stat=instructions:u):
image

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 7]
```diff
@@ -184,16 +187,26 @@ impl<'hir> Map<'hir> {
 
     #[inline]
     pub fn opt_local_def_id(&self, hir_id: HirId) -> Option<LocalDefId> {
-        self.tcx.definitions.opt_hir_id_to_local_def_id(hir_id)
+        // Create a dependency to the owner to ensure the query gets re-executed when the amount of
+        // children changes.
+        self.tcx.ensure().hir_owner_nodes(hir_id.owner);
```
Comment (bjorn3) [inline compiler/rustc_middle/src/hir/map/mod.rs:192]: I think these are the most likely to have caused the perf regression.

File: compiler/rustc_hir/src/arena.rs [hunk 8]
```diff
@@ -43,6 +43,7 @@ macro_rules! arena_types {
             [] stmt: rustc_hir::Stmt<$tcx>,
             [] field_def: rustc_hir::FieldDef<$tcx>,
             [] trait_item_ref: rustc_hir::TraitItemRef,
+            [] trait_candidate: rustc_hir::TraitCandidate,
```
Comment (bjorn3) [inline compiler/rustc_hir/src/arena.rs:46]: This doesn't seem to be used.

**Pull request merged**